### PR TITLE
Proposed Internal Datadog Module Component

### PR DESCRIPTION
**Pull request by gord02**
Component(s):
processor/datadog

The Datadog processor is a component the needs to be deprecated in favor of the Datadog connector. The migration from one component to the other will take effect after two minor releases. Therefore, during the deprecation process, code that is used by the Datadog processor and connector can be shared instead of duplicated.

[Link to relevant issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19740)

### CODE REVIEW

**Comment by atoulme** [comment]
We are going to need codeowners for that new package. Please copy this file for example, and enter codeowners for this path:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/docker/metadata.yaml

Note codeowners _must_ be members of the OpenTelemetry Github organization. More info in CONTRIBUTING.md. Thanks!

File: cmd/otelcontribcol/components.go [hunk 1]
```diff
@@ -4,26 +4,20 @@ package main
 
 import (
 	"go.opentelemetry.io/collector/connector"
-	forwardconnector "go.opentelemetry.io/collector/connector/forwardconnector"
```
Comment (mx-psi) [inline cmd/otelcontribcol/components.go:7]: This file is autogenerated and should have no changes of this kind, could you revert the changes to `components.go` and `components_test.go`?

File: cmd/otelcontribcol/main.go [hunk 2]
```diff
@@ -5,7 +5,6 @@ package main
 
 import (
 	"log"
-
```
Comment (mx-psi) [inline cmd/otelcontribcol/main.go:8]: This file is autogenerated and should not have changes of this type, can you revert the changes on `main.go` and in `main_windows.go`?

File: internal/datadog/go.mod [hunk 3]
```diff
@@ -0,0 +1,72 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog
+
+go 1.19
+
+require (
+	github.com/DataDog/datadog-agent/pkg/trace v0.46.0
```
Comment (mx-psi) [inline internal/datadog/go.mod:6]: This should match the version used here https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/19f2a53e42f9ace56e14a624fdaa56a482082cb8/exporter/datadogexporter/go.mod#L7 can you bump the version?

**Comment by mx-psi** [comment]
@gbbr can you review?

File: cmd/otelcontribcol/components.go [hunk 4]
```diff
@@ -5,7 +5,15 @@ package main
...
 	forwardconnector "go.opentelemetry.io/collector/connector/forwardconnector"
+<<<<<<< HEAD
+	countconnector "github.com/open-telemetry/opentelemetry-collector-contrib/connector/countconnector"
+	exceptionsconnector "github.com/open-telemetry/opentelemetry-collector-contrib/connector/exceptionsconnector"
+	routingconnector "github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector"
+	servicegraphconnector "github.com/open-telemetry/opentelemetry-collector-contrib/connector/servicegraphconnector"
+	spanmetricsconnector "github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector"
+=======
 	"go.opentelemetry.io/collector/exporter"
+>>>>>>> parent of 4fa8eee297 (resolving issues with go dependencies for internal datadog module)
```
Comment (mx-psi) [inline cmd/otelcontribcol/components.go:16]: There are some stray merge conflicts here

File: internal/datadog/go.mod [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/metrics v0.5.2
+	github.com/stretchr/testify v1.8.4
+	go.opentelemetry.io/collector/pdata v1.0.0-rcv0013
```
Comment (mx-psi) [inline internal/datadog/go.mod:9]: Needs a bump
```suggestion
	go.opentelemetry.io/collector/pdata v1.0.0-rcv0014
```

and `make gotidy` afterwards

File: internal/datadog/go.mod (hunk 3 above)
Comment (mx-psi) [reply internal/datadog/go.mod:6]: This still needs addressing

File: cmd/configschema/go.mod [hunk 6]
```diff
@@ -506,6 +506,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/proxy v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/common v0.82.0 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog v0.81.0 // indirect
```
Comment (songy23) [inline cmd/configschema/go.mod:509]: This should be v0.82.0

File: cmd/otelcontribcol/go.mod [hunk 7]
```diff
@@ -515,6 +515,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/metrics v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/proxy v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray v0.82.0 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog v0.81.0 // indirect
```
Comment (songy23) [inline cmd/otelcontribcol/go.mod:518]: Same

File: go.mod [hunk 8]
```diff
@@ -496,6 +496,7 @@ require (
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/xray v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/common v0.82.0 // indirect
 	github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal v0.82.0 // indirect
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog v0.81.0 // indirect
```
Comment (songy23) [inline go.mod:499]: Replace all occurrence of v0.81.0

File: internal/datadog/go.mod (hunk 3 above)
Comment (songy23) [reply internal/datadog/go.mod:6]: I believe this is causing the CI failures.

File: internal/datadog/go.mod (hunk 5 above)
Comment (songy23) [reply internal/datadog/go.mod:9]: This isn't resolved?

File: internal/datadog/go.mod [hunk 9]
```diff
@@ -0,0 +1,72 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/internal/datadog
+
+go 1.19
+
+require (
+	github.com/DataDog/datadog-agent/pkg/trace v0.48.0-devel
+	github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/metrics v0.5.2
```
Comment (mx-psi) [inline internal/datadog/go.mod:7]: If this gets merged after the PR @songy23 is working on we should bump these appropriately

**Review by mx-psi — APPROVED** [review]

**Review by songy23 — COMMENTED** [review]
Is this ready for review or still in draft?

**Comment by gord02** [comment]
@songy23 There are still some checks the PR is not passing but other than that it is ready.

**Comment by songy23** [comment]
> dependabot.yml is out of date, please run "make gendependabot" and commit the changes in this PR.

**Review by songy23 — APPROVED** [review]
LGTM

**Comment by mx-psi** [comment]
needs a rebase

**Review by songy23 — COMMENTED** [review]
> dependabot.yml is out of date, please run "make gendependabot" and commit the changes in this PR.

File: internal/datadog/agent.go [hunk 10]
```diff
@@ -132,7 +132,7 @@ func (p *traceagent) goDrain() {
 
 // Ingest processes the given spans within the traceagent and outputs stats through the output channel
 // provided to newAgent. Do not call Ingest on an unstarted or stopped traceagent.
-func (p *traceagent) Ingest(ctx context.Context, traces ptrace.Traces) {
+func (p *Traceagent) Ingest(ctx context.Context, traces ptrace.Traces) {
```
Comment (gbbr) [inline internal/datadog/agent.go:135]: nit: it would make sense to capitalise the "a" too (TraceAgent), if you did the the "T"

**Review by gbbr — APPROVED** [review]
👍🏻

**Comment by mx-psi** [comment]
> Generated code is out of date, please run "make gengithub" and commit the changes in this PR.

File: .github/dependabot.yml [hunk 11]
```diff
@@ -1092,13 +1102,3 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
-  - package-ecosystem: "gomod"
```
Comment (songy23) [inline .github/dependabot.yml:1095]: Put back line 1095 - 1099

**Review by songy23 — COMMENTED** [review]
I think you probably ran into https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19410, for now please make some manual updates to .github/dependabot.yml to make CI pass

File: .github/dependabot.yml [hunk 12]
```diff
@@ -47,11 +47,7 @@ updates:
...
-    directory: "/connector/datadogconnector"
-    schedule:
-      interval: "weekly"
-      day: "wednesday"
+
```
Comment (songy23) [inline .github/dependabot.yml:50]: Remove the blank line

File: .github/dependabot.yml [hunk 13]
```diff
@@ -1102,3 +1098,14 @@ updates:
     schedule:
       interval: "weekly"
       day: "wednesday"
+
```
Comment (songy23) [inline .github/dependabot.yml:1101]: Remove the blank line

File: .github/dependabot.yml [hunk 14]
```diff
@@ -1102,3 +1098,14 @@ updates:
...
+  - package-ecosystem: "gomod"
+    directory: "/receiver/splunkhecreceiver"
+    schedule:
+      interval: "weekly"
+      day: "wednesday"
+  - package-ecosystem: "gomod"
```
Comment (songy23) [inline .github/dependabot.yml]: Remove line 1107 - 1111

**Review by songy23 — APPROVED** [review]
